pkg/controller/networkpolicy/networkpolicy_controller_test.go

```diff
@@ -15,18 +15,186 @@
+	appliedToGroupStore        storage.Interface
+	addressGroupStore          storage.Interface
+	internalNetworkPolicyStore storage.Interface
+}
+
+func newController() (*fake.Clientset, *networkPolicyController) {
+	client := newClientset()
+	informerFactory := informers.NewSharedInformerFactory(client, informerDefaultResync)
+	appliedToGroupStore := store.NewAppliedToGroupStore()
+	addressGroupStore := store.NewAddressGroupStore()
+	internalNetworkPolicyStore := store.NewNetworkPolicyStore()
+	npController := NewNetworkPolicyController(client, informerFactory.Core().V1().Pods(), informerFactory.Core().V1().Namespaces(), informerFactory.Networking().V1().NetworkPolicies(), addressGroupStore, appliedToGroupStore, internalNetworkPolicyStore)
+	npController.podListerSynced = alwaysReady
+	npController.namespaceListerSynced = alwaysReady
+	npController.networkPolicyListerSynced = alwaysReady
+	return client, &networkPolicyController{
+		npController,
+		informerFactory.Core().V1().Pods().Informer().GetStore(),
+		informerFactory.Core().V1().Namespaces().Informer().GetStore(),
+		informerFactory.Networking().V1().NetworkPolicies().Informer().GetStore(),
+		appliedToGroupStore,
+		addressGroupStore,
+		internalNetworkPolicyStore,
+	}
+}
+
+func newClientset() *fake.Clientset {
+	client := fake.NewSimpleClientset()
+
+	client.PrependReactor("create", "networkpolicies", k8stesting.ReactionFunc(func(action k8stesting.Action) (bool, runtime.Object, error) {
+		np := action.(k8stesting.CreateAction).GetObject().(*networkingv1.NetworkPolicy)
+
+		if np.ObjectMeta.GenerateName != "" {
+			np.ObjectMeta.Name = fmt.Sprintf("%s-%s", np.ObjectMeta.GenerateName, rand.String(8))
+			np.ObjectMeta.GenerateName = ""
+		}
+
+		return false, np, nil
+	}))
+
+	return client
+}
+
+func TestAddNetworkPolicyNoRules(t *testing.T) {
+	ns := metav1.NamespaceDefault
+	npName := "testing-1"
+	pSelector := metav1.LabelSelector{}
+	npObj := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Name: npName, Namespace: ns},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: pSelector,
+		},
+	}
+	apgID := getNormalizedUID(generateNormalizedName(ns, &pSelector, nil))
+	_, npc := newController()
+	npc.addNetworkPolicy(npObj)
+	// An AppliedToGroup must be created as the NetworkPolicy contains a
+	// podSelector.
+	_, found, _ := npc.appliedToGroupStore.Get(apgID)
+	assert.True(t, found, "expected AppliedToGroup to be created")
+	// No AddressGroup must be created as the NetworkPolicy does not contain any
+	// selectors in ingress or egress rules.
+	adgs := npc.addressGroupStore.List()
+	assert.Len(t, adgs, 0, "expected empty AddressGroup list")
+	// An internal NetworkPolicy must be created corresponding to a networkingv1
+	// NetworkPolicy object.
+	key, _ := keyFunc(npObj)
+	_, found, _ = npc.internalNetworkPolicyStore.Get(key)
+	assert.True(t, found, "expected internal NetworkPolicy to be created")
+}
+
+func TestAddNetworkPolicyWithRules(t *testing.T) {
```

Comment:
Would it make sense to duplicate this test and do it once for an ingress rule with a namespace selector and once for an ingress rule with a pod selector?

Reply:
will do it with table style